dali/kernels/reduce/reduce_gpu_impl_test.cu

```diff
@@ -0,0 +1,533 @@
+    { 1, 1, 1 }
+  }};
+  EXPECT_EQ(req.output_shapes[0], ref_out_shape);
+  EXPECT_GE(sum.GetNumStages(), 4);  // both reduced axes must be split due to large max extent
+
+  ScratchpadAllocator sa;
+  sa.Reserve(req.scratch_sizes);
+  auto scratchpad = sa.GetScratchpad();
+  ctx.scratchpad = &scratchpad;
+
+  std::mt19937_64 rng(12345);
+
+  TestTensorList<uint8_t> in;
+  in.reshape(in_shape);
+  auto in_cpu = in.cpu();
+  UniformRandomFill(in_cpu, rng, 0, 255);
+
+  TestTensorList<int64_t> out;
+  out.reshape(req.output_shapes[0]);
+  sum.Run(ctx, out.gpu(), in.gpu());
+
+  auto out_cpu = out.cpu(ctx.gpu.stream);
+  TestTensorList<int64_t> ref;
+  ref.reshape(ref_out_shape);
+  auto ref_cpu = ref.cpu();
+  for (int i = 0; i < ref_cpu.num_samples(); i++) {
+    RefReduce(ref_cpu[i], in_cpu[i], make_span(axes), reductions::sum());
+  }
+
+  Check(out_cpu, ref_cpu);
+}
+
+
+TEST(SumImplGPU, WholeSamples) {
+  TensorListShape<> in_shape = {{
+    { 480, 640 },
+    { 640, 480 },
+    { 0, 0 },
+    { 1280, 300 },
+  }};
+  int axes[] = { 0, 1 };
+  SumImplGPU<int64_t, uint8_t> sum;
+  KernelContext ctx = {};
+  auto req = sum.Setup(ctx, in_shape, make_span(axes), true, false);
+  TensorListShape<> ref_out_shape = {{
+    { 1, 1 },
+    { 1, 1 },
+    { 1, 1 },
+    { 1, 1 }
+  }};
+  EXPECT_EQ(req.output_shapes[0], ref_out_shape);
+  EXPECT_GE(sum.GetNumStages(), 2);
+
+  ScratchpadAllocator sa;
+  sa.Reserve(req.scratch_sizes);
+  auto scratchpad = sa.GetScratchpad();
+  ctx.scratchpad = &scratchpad;
+
+  std::mt19937_64 rng(12345);
+
+  TestTensorList<uint8_t> in;
+  in.reshape(in_shape);
+  auto in_cpu = in.cpu();
+  UniformRandomFill(in_cpu, rng, 0, 255);
+
+  TestTensorList<int64_t> out;
+  out.reshape(req.output_shapes[0]);
+  cudaMemset(out.gpu().data[0], -1, out.gpu().num_elements() * sizeof(int64_t));
+  sum.Run(ctx, out.gpu(), in.gpu());
+
+  auto out_cpu = out.cpu(ctx.gpu.stream);
+  TestTensorList<int64_t> ref;
+  ref.reshape(ref_out_shape);
+  auto ref_cpu = ref.cpu();
+  for (int i = 0; i < ref_cpu.num_samples(); i++) {
+    RefReduce(ref_cpu[i], in_cpu[i], make_span(axes), reductions::sum());
+  }
+
+  Check(out_cpu, ref_cpu);
+}
+
+
+TEST(SumImplGPU, All) {
+  TensorListShape<> in_shape = {{
+    { 480, 640 },
+    { 640, 480 },
+    { 1280, 300 }
+  }};
+  int axes[] = { 0, 1 };
+  SumImplGPU<int64_t, uint8_t> sum;
+  KernelContext ctx = {};
+  auto req = sum.Setup(ctx, in_shape, make_span(axes), true, true);
+  TensorListShape<> ref_out_shape = {{
+    { 1, 1 }
+  }};
+  EXPECT_EQ(req.output_shapes[0], ref_out_shape);
+  EXPECT_GE(sum.GetNumStages(), 2);
+
+  ScratchpadAllocator sa;
+  sa.Reserve(req.scratch_sizes);
+  auto scratchpad = sa.GetScratchpad();
+  ctx.scratchpad = &scratchpad;
+
+  std::mt19937_64 rng(12345);
+
+  TestTensorList<uint8_t> in;
+  in.reshape(in_shape);
+  auto in_cpu = in.cpu();
+  UniformRandomFill(in_cpu, rng, 0, 255);
+
+  TestTensorList<int64_t> out;
+  out.reshape(req.output_shapes[0]);
```

Comment:
Any particular reason why you don't call `cudaMemset` lik in the previous test case?

Reply:
I only used cudaMemset in case of degenerate input (zero size) - in that cases, the expected value is zero, which may coincide with how the memory is initialized on allocation. Here, I have some real value to compare against and it's exceedingly unlikely that the memory will happen to contain the expected result.

Reply:
OK